### ISSUE TITLE
Allow registerForActivityResult() be able to register multiple controllers

### ISSUE THREAD

**Issue by dimsuz** [bluelinelabs/Conductor#325]
Current behavior:

```kotlin
class Controller1 {
  init {
    registerForActivityResult(SOME_REQUEST_CODE)
  }
}
// ========
class Controller2 {
  init {
    registerForActivityResult(SOME_REQUEST_CODE)
  }
}
// ========
class Controller3 {
  fun someFunction() {
    startActivityForResult(SOME_REQUEST_CODE, SomeActivity::class.java)
  }
}
```

in this situation only `Controller3` will receive `onActivityResult()` callback invocation.

It is not the first time I get caught up in situation where I expect that former two controllers would be "subscribed" too (documentation of `registerForActivityResult()` doesn't clearly suggest otherwise)

This happens because each request code is internally mapped to only **one** Controller - the latest one:

```java
    public void registerForActivityResult(@NonNull String instanceId, int requestCode) {
        activityRequestMap.put(requestCode, instanceId);
    }
```

Do you think that this can be extended so that a list of controller `instanceId`s would be held there?

It would seriously simplify some situations, where several controllers must react to some external activity result. Otherwise one would have to use a similar mechanism where the controller which started that activity propagates a result to several controllers (`setTargetController`?)

Or do you see this as a bad pattern? Or maybe it would fit badly in library design?
Objections or suggestions are welcome. If you have nothing against this, I could submit a PR.

**Comment by shakil807g** [comment]
i have a similar problem  i call this call from a controller  which  picker image from gallery 


```
`  imagePicker = new ImagePicker(getActivity());
        imagePicker.shouldGenerateMetadata(true);
        imagePicker.shouldGenerateThumbnails(true);
        imagePicker.setImagePickerCallback(new ImagePickerCallback() {
            @Override
            public void onImagesChosen(List<ChosenImage> list) {
                Log.d(TAG, "onImagesChosen: "+list.get(0).getOriginalPath());
                txtMessage.setText(list.get(0).getOriginalPath());
            }

            @Override
            public void onError(String s) {

            }
        });
        imagePicker.pickImage();`
```

and i register onActivityResult in controller like this 

 ```
@Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        Log.d(TAG, "onActivityResult: ");

        if (resultCode == Activity.RESULT_OK) {
            if (requestCode == Picker.PICK_IMAGE_DEVICE) {
                if (imagePicker == null) {
                    imagePicker = new ImagePicker(getActivity());
                }
                imagePicker.submit(data);
            }


        }
    }
```

but it never get called but  onActivityResult   on my MainActivity get called  i think it is due to iam passing activity context  but i was expecting onActivityResult  to be called on controller as well

**Comment by dimsuz** [comment]
How do you start your activity inside image picker class? For your Controller.onActivityResult to be called you must use Controller.startActivityForResult, not Activity.startActivityForResult

**Comment by dimsuz** [comment]
Or you must manually call Controller.registerForActivityResult if you are calling Activity.startActivityForResult and not controller's

**Comment by shakil807g** [comment]
yes ur right but that the problem is  ImagePicker  is from external library and it is calling Activity.startActivityForResult  how can  pass the result from activity,s onResult  to that controller  or may there is some better way ???

**Comment by shakil807g** [comment]
i think it will be easier if  Activity,s onResult  will propagate to all the controller and each controller can handle Activity,s onResult as well

**Comment by dimsuz** [comment]
As I've said in a latter comment, you should just use Controller.registerForActivityResult before calling pickImage, this should then propagate all results to your Controller

**Comment by shakil807g** [comment]
oh thanks but should i have to call registerForActivityResult for multiple result

**Comment by dimsuz** [comment]
Oh, you mean ImagePicker supports several *request codes*? Yes, you just call it several times providing your request codes to which you need to register. But I see only one: Picker.PICK_IMAGE_DEVICE

**Comment by dimsuz** [comment]
This issue describes another problem: you can't register multiple controllers to one request code

**Comment by shakil807g** [comment]
i am calling ImagePicker like this

```
registerForActivityResult(PICK_IMAGE_DEVICE);
        imagePicker = new ImagePicker(getActivity());
        imagePicker.shouldGenerateMetadata(true);
        imagePicker.shouldGenerateThumbnails(true);
        imagePicker.setImagePickerCallback(new ImagePickerCallback() {
            @Override
            public void onImagesChosen(List<ChosenImage> list) {
                Log.d(TAG, "onImagesChosen: "+list.get(0).getOriginalPath());
                txtMessage.setText(list.get(0).getOriginalPath());
            }


            @Override
            public void onError(String s) {

            }
        });
        imagePicker.pickImage();
```
but still it is not calling  controller onActivityResult

**Comment by shakil807g** [comment]
i think switching this to library as an alternative to fragment will bring it own problems

**Comment by PaulWoitaschek** [comment]
> i think switching this to library as an alternative to fragment will bring it own problems

Absolutely. Everything in software development world brings its own problems.
